external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java

```diff
@@ -161,13 +153,19 @@ public V get(K key, V defaultValue) {
     @Override
     public V delete(K key) {
         LOG.debug("delete key '{}'", key);
-        String redisKey = encode(keySerializer.serialize(key));
+        String redisKey = encoder.encodeKey(key);
         V curr = get(key);
-        pendingPrepare.put(redisKey, TOMBSTONE);
+        pendingPrepare.put(redisKey, RedisEncoder.TOMBSTONE);
         return curr;
     }
 
     @Override
+    public Iterator<Map.Entry<K, V>> iterator() {
+        return new RedisKeyValueStateIterator<K, V>(namespace, jedisContainer, 
```

Comment:
You also need to consider the entries in `pendingPrepare` and `pendingCommit`. See the `get` api for example. `RedisKeyValueStateIterator` should iterate over pendingPrepare and pendingCommit before iterating over the entries in Redis.